File: check-for-unread.py
class CBOR:

    @staticmethod
    def _error(message):
        raise Exception(message)
    
    @staticmethod
    def _is_int(value):
        if not isinstance(value, int): CBOR._error("'int' expected")
        return value

    class _CborObject(object):
        def __init__(self):
            self._read_flag = False

        # Common methods
        def check_for_unread(self):
            # Top-level Array, Map, and Tag container object marked as read.
            self._mark_as_read(self)
            self._traverse(None, None)

        # Private methods
        def _mark_as_read(self, object):
            if not object._is_primitive():
                object._read_flag = True
            return object
        
        def _is_primitive(self):
            return isinstance(self, CBOR.Int) or isinstance(self, CBOR.String)

        def _traverse(self, holding_object, map_key):
            match type(self).__name__:
                case "Map":
                    for entry in self._entries: 
                        entry.value._traverse(self, entry.key)
                case "Array":
                    for object in self._objects: object._traverse(self, None)
                case "Tag":
                    self._object._traverse(self, None)
            if not self._read_flag:
                problem_item = type(self).__name__
                if self._is_primitive():
                    problem_item += " with value={}".format(
                        self.value if isinstance(self, CBOR.Int) else self.string)
                problem_item += " was never read"
                if holding_object is not None:
                    if isinstance(holding_object, CBOR.Array):
                        holder = "Array element of type"
                    elif isinstance(holding_object, CBOR.Tag):
                        holder = "Tagged object {} of type".format(
                            holding_object._tag_number)
                    else:
                        holder = "Map key {} with argument".format(map_key)
                    problem_item = holder + " " + problem_item
                CBOR._error(problem_item)

    class Map(_CborObject):
        def __init__(self):
            super().__init__()
            self._entries = list()

        def set(self, key, object):
            self._entries.append(CBOR.Map.Entry(CBOR._is_int(key), object))
            return self

        def get(self, key):
            CBOR._is_int(key)
            for entry in self._entries:
                if entry.key == key:
                    return self._mark_as_read(entry.value)
            CBOR._error("Key {0} not found".format(key))

        class Entry:
            def __init__(self, key, value):
                self.key = key
                self.value = value

    class Array(_CborObject):
        def __init__(self):
            super().__init__()
            self._objects = list()

        def add(self, object):
            self._objects.append(object)
            return self

        def get(self, index):
            return self._mark_as_read(self._objects[CBOR._is_int(index)])
                
    class Tag(_CborObject):
        def __init__(self, tag_number, object):
            super().__init__()
            self._tag_number = tag_number
            self._object = object

        def get(self):
            return self._mark_as_read(self._object)

    class String(_CborObject):
        def __init__(self, string):
            super().__init__()
            if not isinstance(string, str):
                CBOR._error("'str' type expected")
            self.string = string

        def get_string(self):
            self._read_flag = True
            return self.string
        
    class Int(_CborObject):
        def __init__(self, value):
            super().__init__()
            if not isinstance(value, int):
                CBOR._error("'int' type expected")
            self.value = value

        def get_int(self):
            self._read_flag = True
            return self.value


def test(statement, access, message=None):
    # print(":" + statement)
    fail = message is not None
    error = ""
    res = eval(statement)
    try:
        res.check_for_unread()
        assert access is None and not fail, statement
    except Exception as e:
        error = repr(e)
        # print(statement + " " + error)
        assert access or fail, statement
        assert error.find("never read") > 0
    if access:
        try:
            eval("res." + access)
            res.check_for_unread()
            assert not fail, statement
        except Exception as e:
            error = repr(e)
            # print(statement + " " + error)
            assert fail, statement
            assert error.find("never read") > 0
    if message is not None and (error.find(message) < 0):
        assert False, "not" + repr(message) + error


test("CBOR.Array()", None)
test("CBOR.Map()", None)
test("CBOR.Tag(45, CBOR.Map())", "get()")

test("CBOR.Tag(45, CBOR.Map().set(1, CBOR.Int(6)))", "get().get(1)",
     "Map key 1 with argument Int with value=6 was never read")

test("CBOR.Tag(45, CBOR.Map().set(1, CBOR.Int(6)))", "get().get(1).get_int()")
test("CBOR.Array().add(CBOR.Tag(45, CBOR.Map()))", "get(0)",
     "Tagged object 45 of type Map was never read")

test("CBOR.Array().add(CBOR.Tag(45, CBOR.Map()))", "get(0).get()")

test("CBOR.Array().add(CBOR.Tag(45, CBOR.Int(6)))", "get(0).get()",
     "Tagged object 45 of type Int with value=6 was never read")

test("CBOR.Array().add(CBOR.Tag(45, CBOR.Int(6)))", "get(0).get().get_int()")

test("CBOR.Array().add(CBOR.String('Hi!'))", "get(0)",
     "Array element of type String with value=Hi! was never read")

test("CBOR.Array().add(CBOR.Int(6))", "get(0).get_int()")

test("CBOR.Map().set(1, CBOR.Array())", None,
     "Map key 1 with argument Array was never read")

test("CBOR.Map().set(1, CBOR.Array())", "get(1)")

test("CBOR.Tag(45, CBOR.Map().set(1, CBOR.Int(6)))", "get().get(1).get_int()")

test("CBOR.Array().add(CBOR.Array())", "get(0)")

test("CBOR.Array().add(CBOR.Array())", None,
     "Array element of type Array was never read")

test("CBOR.Int(6)", "get_int()")

# a slightly more elaborate example
res = CBOR.Map().set(2, CBOR.Array()).set(1, CBOR.String("Hi!"))
res.get(2).add(CBOR.Int(700))
assert res.get(2).get(0).get_int() == 700
assert res.get(1).get_string() == "Hi!"
res.check_for_unread() # all is good

File: check-for-unread.js
class CBOR {

    static #error(message) {
        throw new Error(message);
    }
    
    static #isInt(value) {
        if (!Number.isSafeInteger(value)) CBOR.#error("'integer' expected");
        return value;
    }

    static #typeCheck(object, expected) {
        if (typeof object == expected) {
            return object;
        }
        CBOR.#error("'" + expected + "' expected ");
    }

    static CborObject = class {
        constructor() {
            this._readFlag = false;
        }

        // Common methods
        checkForUnread() {
            this._markAsRead(this);
            this.#traverse(null);
        }

        // Private methods
        _markAsRead(object) {
            if (!this._is_primitive(object)) {
                object._readFlag = true;
            }
            return object;
        }

        _is_primitive(object) {
            return object instanceof CBOR.Int || object instanceof CBOR.String;
        }

        #traverse(holding_object) {
            switch (this.constructor.name) {
                case "Map":
                    this._entries.forEach(entry => {
                        entry.value.#traverse(entry.key);
                    });
                    break;
                case "Array":
                    this._objects.forEach(object => {
                         object.#traverse(this);
                    });
                    break;
                case "Tag":
                    this._object.#traverse(this);
                    break;
            }
            if (!this._readFlag) {
                let problem_item = this.constructor.name;
                if (this._is_primitive(this)) { 
                    problem_item += " with value=" + this.value;
                }
                problem_item += " was never read";
                let holder;
                if (holding_object) {
                    if (holding_object instanceof CBOR.Array) {
                        holder = "Array element of type";
                    } else if (holding_object instanceof CBOR.Tag) {
                        holder = "Tagged object " + holding_object._tag_number + " of type";
                    } else {
                        holder = "Map key " + holding_object + " with argument";
                    }
                    problem_item = holder + " " + problem_item;
                }
                CBOR.#error(problem_item);
            }
        }
    }

    static Map = class extends CBOR.CborObject {
        constructor() {
            super();
            this._entries = [];
        }

        set(key, value) {
            this._entries.push(new CBOR.Map.Entry(CBOR.#isInt(key), value));
            return this;
        }

        get(key) {
            CBOR.#isInt(key);
            for (let i = 0; i < this._entries.length; i++) {
                if (this._entries[i].key == key) {
                    return this._markAsRead(this._entries[i].value);
                }
            }
            CBOR.#error("Key " + key + " not found");
        }

        static Entry = class {
            constructor(key, value) {
                this.key = key;
                this.value = value;
            }
        }
    }

    static Array = class extends CBOR.CborObject {
        constructor() {
            super();
            this._objects = [];
        }

        add(object) {
            this._objects.push(object);
            return this;
        }

        get(index) {
            return this._markAsRead(this._objects[CBOR.#isInt(index)]);
        }
    }
                
    static Tag = class extends CBOR.CborObject {
        constructor(tag_number, object) {
            super();
            this._tag_number = tag_number;
            this._object = object;
        }

        get() {
            return this._markAsRead(this._object);
        }
    }

    static String = class extends CBOR.CborObject {
        constructor(string) {
            super();
            this.string = CBOR.#typeCheck(string, "string");
        }

        getString() {
            this._readFlag = true;
            return this.string;
        }
    }
        
    static Int = class extends CBOR.CborObject {
        constructor(value) {
            super();
            this.value = CBOR.#isInt(value);
        }

        getInt() {
            this._readFlag = true;
            return this.value;
        }
    }

    // The Proxy concept enables checks for invocation by "new" and number of arguments.
    static #handler = class {

        constructor(numberOfArguments) {
            this.numberOfArguments = numberOfArguments;
        }

        apply(target, thisArg, argumentsList) {
            if (argumentsList.length != this.numberOfArguments) {
                CBOR.#error("CBOR." + target.name + " expects " + 
                    this.numberOfArguments + " argument(s)");
            }
            return new target(...argumentsList);
        }

        construct(target, args) {
            CBOR.#error("CBOR." + target.name + " does not permit \"new\"");
        }
    }

    static Array = new Proxy(CBOR.Array, new CBOR.#handler(0));
    static Map = new Proxy(CBOR.Map, new CBOR.#handler(0));
    static Tag = new Proxy(CBOR.Tag, new CBOR.#handler(2));
    static Int = new Proxy(CBOR.Int, new CBOR.#handler(1));
    static String = new Proxy(CBOR.String, new CBOR.#handler(1));
}

//////////////////////////////
//          TEST            //
//////////////////////////////

function assertTrue(expression, explanation) {
    if (!expression) {
        throw new Error(explanation);
    }
}

function test(statement, access, message) {
    // console.log(":" + statement);
    let fail = message != null;
    let error = "";
    let res = eval(statement);
    try {
        res.checkForUnread();
        assertTrue(!access && !fail, statement);
    } catch (e) {
        error = e.toString();
        // console.log(statement + " " + error);
        assertTrue(access || fail, statement);
        assertTrue(error.includes("never read"));
    }
    if (access) {
        try {
            eval("res." + access)
            res.checkForUnread();
            assertTrue(!fail, statement);
        } catch (e) {
            error = e.toString();
            // console.log(statement + " " + error)
            assertTrue(fail, statement);
            assertTrue(error.includes("never read"));
        }
    }
    if (!message && error.includes(message)) {
        assertTrue(false, "not" + message + error);
    }
}

test("CBOR.Array()", null)
test("CBOR.Map()", null)
test("CBOR.Tag(45, CBOR.Map())", "get()")

test("CBOR.Tag(45, CBOR.Map().set(1, CBOR.Int(6)))", "get().get(1)",
     "Map key 1 with argument Int with value=6 was never read")

test("CBOR.Tag(45, CBOR.Map().set(1, CBOR.Int(6)))", "get().get(1).getInt()")
test("CBOR.Array().add(CBOR.Tag(45, CBOR.Map()))", "get(0)",
     "Tagged object 45 of type Map was never read")

test("CBOR.Array().add(CBOR.Tag(45, CBOR.Map()))", "get(0).get()")

test("CBOR.Array().add(CBOR.Tag(45, CBOR.Int(6)))", "get(0).get()",
     "Tagged object 45 of type Int with value=6 was never read")

test("CBOR.Array().add(CBOR.Tag(45, CBOR.Int(6)))", "get(0).get().getInt()")

test("CBOR.Array().add(CBOR.String('Hi!'))", "get(0)",
     "Array element of type String with value=Hi! was never read")

test("CBOR.Array().add(CBOR.Int(6))", "get(0).getInt()")

test("CBOR.Map().set(1, CBOR.Array())", null,
     "Map key 1 with argument Array was never read")

test("CBOR.Map().set(1, CBOR.Array())", "get(1)")

test("CBOR.Tag(45, CBOR.Map().set(1, CBOR.Int(6)))", "get().get(1).getInt()")

test("CBOR.Array().add(CBOR.Array())", "get(0)")

test("CBOR.Array().add(CBOR.Array())", null,
     "Array element of type Array was never read")

test("CBOR.Int(6)", "getInt()")

// a slightly more elaborate example
res = CBOR.Map().set(2, CBOR.Array()).set(1, CBOR.String("Hi!"))
res.get(2).add(CBOR.Int(700))
assertTrue(res.get(2).get(0).getInt() == 700)
assertTrue(res.get(1).getString() == "Hi!")
res.checkForUnread() // all is good
